opentimelineio_contrib/application_plugins/rv/example_otio_reader/otio_reader.py

```diff
@@ -0,0 +1,263 @@
+        in_seq
+    )
+
+    track_kind = in_seq.kind
+
+    new_inputs = []
+    for thing in items_to_serialize:
+        if isinstance(thing, tuple):
+            result = _create_transition(*thing, track_kind=track_kind)
+        elif thing.duration().value == 0:
+            continue
+        else:
+            result = create_rv_node_from_otio(thing, track_kind)
+
+        if result:
+            new_inputs.append(result)
+
+    commands.setNodeInputs(new_seq, new_inputs)
+    return new_seq
+
+
+def _create_timeline(tl, _=None):
+    return create_rv_node_from_otio(tl.tracks)
+
+
+def _create_media_reference(item, track_kind=None):
+    if hasattr(item, "media_reference") and item.media_reference:
+        if isinstance(item.media_reference, otio.schema.ExternalReference):
+            media = [str(item.media_reference.target_url)]
+            if track_kind == otio.schema.TrackKind.Audio:
+                # Create blank video media to accompany audio for valid source
+                blank = "{},start={},end={},fps={}.movieproc".format(
+                    "blank",
+                    item.available_range().start_time.value,
+                    item.available_range().end_time_inclusive().value,
+                    item.available_range().duration.rate
+                )
+                # Appending blank to media promotes name of audio file in RV
+                media.append(blank)
+
+            return media
+        elif isinstance(item.media_reference, otio.schema.GeneratorReference):
+            if item.media_reference.generator_kind == "SMPTEBars":
+                kind = "smptebars"
+                return (
+                    [
+                        "{},start={},end={},fps={}.movieproc".format(
+                            kind,
+                            item.available_range().start_time.value,
+                            item.available_range().end_time_inclusive().value,
+                            item.available_range().duration.rate
+                        )
+                    ]
+                )
+    return None
+
+
+def _create_item(it, track_kind=None):
+    range_to_read = it.trimmed_range()
+
+    if not range_to_read:
+        raise otio.exceptions.OTIOError(
+            "No valid range on clip: {0}.".format(
+                str(it)
+            )
+        )
+
+    new_media = _create_media_reference(it, track_kind)
+    if not new_media:
+        kind = "smptebars"
+        if isinstance(it, otio.schema.Gap):
+            kind = "blank"
+        new_media = [
+                        "{},start={},end={},fps={}.movieproc".format(
+                            kind,
+                            range_to_read.start_time.value,
+                            range_to_read.end_time_inclusive().value,
+                            range_to_read.duration.rate
+                        )
+                    ]
+
+    src = commands.addSourceVerbose(new_media)
```

Comment:
Is there possibly a way to add the Sources outside of this adapter and, if they've already been added, have this adapter connect the existing sources into the Sequence?

I ask because, in our experience, while addSourceVerbose does add the source (verbosely), but it doesn't actively update the UI, so if you have several sources in your OTIO (as one would expect for a feature sequence), the user doesn't get any feedback until all the sources have been added. Maybe there's a world where you add all the sources for the OTIO in one step using rv.commands.addSources() (which provides a "Loading X of X" dialog) and then assemble the Sequence from the existing sources. (rv.commands.addSources() fires after-progressive-loading after it completes, so this second step could take place there.)

For an example OTIO reader, that may be overkill, but for production, it's something to consider.

Have you guys tried anything like that?

Reply:
I've tried using QtGui.QApplication.processEvents() to solve this problem, but the nature of processEvents can reverse the order of items loaded and cause other unintended consequences.  You get a nice live loading feel though.  Maybe a more specific processEvents command could be a solution.  

I'm not sure which events to include though. IE
app.processEvents(QtCore.QEventLoop.ExcludeUserInputEvents, stepSize)